human-viz-bridge/src/HumanJointStateBridge.cpp

```diff
@@ -0,0 +1,188 @@
+
+#include <iDynTree/Core/TestUtils.h>
+#include <iDynTree/ModelIO/ModelLoader.h>
+#include <TickTime.h>
+#include <yarp/os/BufferedPort.h>
+#include <yarp/os/LogStream.h>
+#include <yarp/os/Network.h>
+#include <yarp/os/Node.h>
+#include <yarp/os/Publisher.h>
+#include <yarp/os/RFModule.h>
+#include <yarp/os/Time.h>
+
+#include <cmath>
+#include <iostream>
+#include <limits.h>
+#include <string>
+#include <vector>
+
+using namespace std;
+using namespace yarp::os;
+using namespace human;
+
+inline TickTime normalizeSecNSec(double yarpTimeStamp)
+{
+    uint64_t time = (uint64_t) (yarpTimeStamp * 1000000000UL);
+    uint64_t nsec_part = (time % 1000000000UL);
+    uint64_t sec_part = (time / 1000000000UL);
+    TickTime ret;
+
+    if (sec_part > UINT_MAX)
+    {
+        yWarning() << "Timestamp exceeded the 64 bit representation, resetting it to 0";
+        sec_part = 0;
+    }
+
+    ret.sec  = sec_part;
+    ret.nsec = nsec_part;
+    return ret;
+}
+
+class xsensJointStatePublisherModule : public RFModule, public TypedReaderCallback<HumanState> 
+{
+    Port client_port;
+    BufferedPort<HumanState> humanStateDataPort;
+
+    Publisher<sensor_msgs_JointState> publisher;
+    Publisher<tf2_msgs_TFMessage> publisher_tf;
+    
+    //variables declaration
+    sensor_msgs_JointState joint_state;
+    tf2_msgs_TFMessage tf;
+    
+public:
+    double getPeriod()
+    {
+        // module periodicity (seconds).
+        return 100;
+    }
+    // Main function. Will be called periodically every getPeriod() seconds
+    bool updateModule()
+    {
+        return true;
+    }
+    // Configure function. 
+    bool configure(ResourceFinder &rf)
+    {
+        string moduleName = rf.check("name", Value("human-jointstate-bridge"), "Checking module name").asString();
+        string serverName = rf.check("serverName", Value("human-state-provider"), "Checking server name").asString();
+        string urdfModelFile = rf.findFile("urdf_model");
+        setName(moduleName.c_str());
+        
+        // Load model
+        iDynTree::ModelLoader modelLoader;
+        iDynTree::Model model;
+        if (!modelLoader.loadModelFromFile(urdfModelFile)) {
+            yError() << "Failed to load URDF model from file " << urdfModelFile;
+            return false;
+        }
+        
+        model = modelLoader.model();
+
+        if (!publisher.topic("/human/joint_states")) {
+            cerr<< "Failed to create publisher to /joint_states\n";
+            return false;
+        }
+        
+        if (!publisher_tf.topic("/tf")) {
+        cerr<< "Failed to create publisher to /tf\n";
+        return -1;
+        }
+        
+        tf.transforms.resize(1);
+        tf.transforms[0].header.frame_id = "ground";
+        tf.transforms[0].child_frame_id = "Pelvis";
+        
+        vector<string> joints;
+        joints.reserve(model.getNrOfJoints());
+        for (iDynTree::JointIndex jointIndex = 0; jointIndex < model.getNrOfJoints(); ++jointIndex) {
```

Comment:
Be careful that the order of the joints passed from `human-state-provider` can be different